sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala

```diff
@@ -0,0 +1,210 @@
+ *   - Non-index columns are cast to a common data type, determined by finding the least
+ *     common type that can accommodate all non-index columns.
+ *   - The data is sorted by the index column, and rows with `null` index values are excluded
+ *     from the transpose operation.
+ *   - The transposed DataFrame is constructed by turning the original rows into columns, with
+ *     the index column values becoming the new column names and the non-index column values
+ *     populating the transposed data.
+ */
+class ResolveTranspose(sparkSession: SparkSession) extends Rule[LogicalPlan] {
+
+  private def leastCommonType(dataTypes: Seq[DataType]): DataType = {
+    if (dataTypes.isEmpty) {
+      StringType
+    } else {
+      dataTypes.reduce { (dt1, dt2) =>
+        TypeCoercion.findTightestCommonType(dt1, dt2).getOrElse {
+          throw new AnalysisException(
+            errorClass = "TRANSPOSE_NO_LEAST_COMMON_TYPE",
+            messageParameters = Map(
+              "dt1" -> dt1.toString,
+              "dt2" -> dt2.toString)
+          )
+        }
+      }
+    }
+  }
+
+  private def transposeMatrix(
+      fullCollectedRows: Array[InternalRow],
+      nonIndexColumnNames: Seq[String],
+      nonIndexColumnDataTypes: Seq[DataType]): Array[Array[Any]] = {
+    val numTransposedRows = fullCollectedRows.head.numFields - 1
+    val numTransposedCols = fullCollectedRows.length + 1
+    val finalMatrix = Array.ofDim[Any](numTransposedRows, numTransposedCols)
+
+    // Example of the original DataFrame:
+    // +---+-----+-----+
+    // | id|col1 |col2 |
+    // +---+-----+-----+
+    // |  1|  10 |  20 |
+    // |  2|  30 |  40 |
+    // +---+-----+-----+
+    //
+    // After transposition, the finalMatrix will look like:
+    // [
+    //   ["col1", 10, 30],  // Transposed row for col1
+    //   ["col2", 20, 40]   // Transposed row for col2
+    // ]
+
+    for (i <- 0 until numTransposedRows) {
+      // Insert non-index column name as the first element in each transposed row
+      finalMatrix(i)(0) = UTF8String.fromString(nonIndexColumnNames(i))
+
+      for (j <- 1 until numTransposedCols) {
+        // Insert the transposed data
+
+        // Example: If j = 2, then row = fullCollectedRows(1)
+        // This corresponds to the second row of the original DataFrame: InternalRow(2, 30, 40)
+        val row = fullCollectedRows(j - 1)
+
+        // Example: If i = 0 (for "col1"), and j = 2,
+        // then finalMatrix(0)(2) corresponds to row.get(1, nonIndexColumnDataTypes(0)),
+        // which accesses the value 30 from InternalRow(2, 30, 40)
+        finalMatrix(i)(j) = row.get(i + 1, nonIndexColumnDataTypes(i))
+      }
+    }
+    finalMatrix
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+    _.containsPattern(TreePattern.UNRESOLVED_TRANSPOSE)) {
+    case t @ UnresolvedTranspose(indexColumnOpt, child) if child.resolved =>
```

Comment:
shall we also wait for the index column to be resolved first?

Reply:
Makes sense! Adjusted.